cooked-validators/src/Cooked/Tx/Constraints/Type.hs

```diff
@@ -49,108 +51,295 @@ import Type.Reflection
 --
 -- - an 'AT' for affine traversals
 
--- * 'SpendableOut': The type of UTxOs
-
--- | A 'SpendableOut' is an outref that is ready to be spent; with its
---  underlying 'Pl.ChainIndexTxOut'.
-data SpendableOut = SpendableOut
-  { sOutTxOutRef :: Pl.TxOutRef,
-    sOutChainIndexTxOut :: Pl.ChainIndexTxOut
-  }
-  deriving (Eq, Show)
-
-makeLensesFor
-  [ ("sOutTxOutRef", "sOutTxOutRefL"),
-    ("sOutChainIndexTxOut", "sOutChainIndexTxOutL")
-  ]
-  ''SpendableOut
-
-instance Ord SpendableOut where
-  -- TODO: Is this sufficient, i.e. can there be well-formed 'SpendableOut's
-  -- that have the same 'TxOutRef', but different 'ChainIndexTxOut's?
-  (<=) = (<=) `on` sOutTxOutRef
-
-sOutValueL :: Lens' SpendableOut Pl.Value
-sOutValueL = sOutChainIndexTxOutL % lensVL Pl.ciTxOutValue
-
-sOutValue :: SpendableOut -> Pl.Value
-sOutValue = (^. sOutValueL)
-
-sOutAddressL :: Lens' SpendableOut Pl.Address
-sOutAddressL = sOutChainIndexTxOutL % lensVL Pl.ciTxOutAddress
-
-sOutAddress :: SpendableOut -> Pl.Address
-sOutAddress = (^. sOutAddressL)
-
--- | If a 'SpendableOut' belongs to a public key, return its hash.
-sBelongsToPubKey :: SpendableOut -> Maybe Pl.PubKeyHash
-sBelongsToPubKey s = case Pl.addressCredential (s ^. sOutAddressL) of
-  Pl.PubKeyCredential pkh -> Just pkh
+-- * 'IsOutput': UTxOs that can be used as transaction outputs
+
+-- | A generalisation of 'Pl.TxOut': With the two type families, we can lift
+-- some information about who owns the output (a public key, a script...?) and
+-- the datum (do we have an inline datum, a datum hash, nothing...?) to the type
+-- level.
+class Show o => IsOutput o where
+  -- The owner type can be, in particular, a 'TypedValidator a' or a
+  -- 'PubkeyHash'
+  type OwnerType o
+  type DatumType o
+  type ValueType o
+  outputOwnerL :: Lens' o (OwnerType o)
+  outputStakingCredentialL :: Lens' o (Maybe Pl.StakingCredential)
+  outputDatumL :: Lens' o (DatumType o)
+  outputValueL :: Lens' o (ValueType o)
+  outputAddress :: o -> Pl.Address
+  outputOutputDatum :: o -> Pl.OutputDatum
+  outputValue :: o -> Pl.Value
```

Comment:
This type class leaves me a bit skeptical. It contains everything we need but I am a bit afraid of having two levels of abstraction in the same structure. I'm also wondering why the abstract representation of the elements in the output could not internally yield the concrete elements. In other words, I would expect the three last functions (line 70 to 72) to be embedded inside the types of the other functions. For example, if I can retrieve both the OwnerType as well as the staking credential, I would assume to necessarily have access to the plutus address.

Reply:
Yes, this would suit me to have these function taken outside as you did, and regrouped into an adequate typeclass. My only additional suggestion would be to simply call it `IsOnChainOutput` instead of this long `IsOutputWithOnChainRepresentation`